File: monolith-spigot/src/main/java/gg/warcraft/monolith/spigot/item/SpigotInventory.java
package gg.warcraft.monolith.spigot.item;

import com.google.inject.Inject;
import com.google.inject.assistedinject.Assisted;
import gg.warcraft.monolith.api.item.Inventory;
import gg.warcraft.monolith.api.world.item.Item;
import gg.warcraft.monolith.api.world.item.ItemType;
import gg.warcraft.monolith.spigot.world.item.SpigotItemMapper;
import org.bukkit.inventory.ItemStack;

import java.util.*;

public class SpigotInventory implements Inventory {
    private final SpigotItemMapper itemMapper;
    private final org.bukkit.inventory.Inventory inventory;

    @Inject
    public SpigotInventory(SpigotItemMapper itemMapper, @Assisted org.bukkit.inventory.Inventory inventory) {
        this.itemMapper = itemMapper;
        this.inventory = inventory;
    }

    @Override
    public List<Item> getItems() {
        System.out.println("WARNING Attempted to read items from inventory, always returning empty");
        return new ArrayList<>(); // TODO Arrays.stream(inventory.getContents())
//                .filter(Objects::nonNull)
//                .map(itemMapper::map)
//                .collect(Collectors.toList());
    }

    @Override
    public int getSpace() {
        return (int) Arrays.stream(inventory.getStorageContents())
                .filter(Objects::isNull)
                .count();
    }

    @Override
    public List<Item> add(List<Item> items) {
        ItemStack[] spigotItems = items.stream()
                .map(itemMapper::map)
                .toArray(ItemStack[]::new);
        Map<Integer, ItemStack> excessItems = inventory.addItem(spigotItems);
        System.out.println("WARNING Attempted to add item to inventory, always returning empty");
        return new ArrayList<>();
        // TODO return excessItems.values().stream()
//                .map(itemMapper::map)
//                .collect(Collectors.toList());
    }

    @Override
    public void clear() {
        inventory.clear();
    }

    @Override
    public boolean contains(ItemType type) {
        // TODO return inventory.contains(type.getId(), type.getData());
        System.out.println("WARNING Attempted to read if " + type + " is in inventory, always returning false");
        return false;
    }

    @Override
    public boolean contains(Item item) {
        ItemStack spigotItem = itemMapper.map(item);
        return inventory.contains(spigotItem, 1); // TODO item.getStackSize());
    }

    @Override
    public boolean remove(Item item) {
        ItemStack spigotItem = itemMapper.map(item);
        if (inventory.containsAtLeast(spigotItem, 1)) { // TODO item.getStackSize())) {
            inventory.removeItem(spigotItem);
            return true;
        }
        return false;
    }
}

File: monolith-spigot/src/main/scala/gg/warcraft/monolith/spigot/item/SpigotInventory.scala
/*
 * MIT License
 *
 * Copyright (c) 2020 WarCraft <https://github.com/WarCraft>
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

package gg.warcraft.monolith.spigot.item

import gg.warcraft.monolith.api.item.{Inventory, Item, ItemType, ItemVariant}

class SpigotInventory(
    private val inventory: SpigotItemInventory
)(
    implicit private val itemTypeMapper: SpigotItemTypeMapper,
    implicit private val itemVariantMapper: SpigotItemVariantMapper,
    implicit private val itemMapper: SpigotItemMapper
) extends Inventory {
  override lazy val items: List[Item] =
    inventory.getContents
      .map { itemMapper.map(_, None) }
      .filter(_.isDefined)
      .map(_.get)
      .toList

  override def hasSpace(count: Int): Boolean =
    count <= inventory.getStorageContents.count(_ == null)

  // TODO add more elaborate checks for merge-able items
  override def hasSpaceFor(items: Item*): Boolean =
    hasSpace(items.length)

  override def contains(`type`: ItemType, count: Int): Boolean = {
    val material = itemTypeMapper.map(`type`)
    inventory.contains(material, count)
  }

  override def contains(`type`: ItemType): Boolean =
    contains(`type`, count = 1)

  override def contains(variant: ItemVariant, count: Int): Boolean = {
    val material = itemVariantMapper.map(variant)
    inventory.contains(material, count)
  }

  override def contains(variant: ItemVariant): Boolean =
    contains(variant, count = 1)

  override def contains(item: Item): Boolean = {
    val spigotItem = itemMapper.map(item)
    inventory.contains(spigotItem, spigotItem.getAmount)
  }
}
